README.md
<p align="center">
  <img src="https://hsto.org/webt/ek/0q/co/ek0qcotut4ioaifdup8nroyko2a.png" alt="logo" width="128" />
</p>

# `go-filecache`

![Release version][badge_release_version]
![Project language][badge_language]
[![Build Status][badge_build]][link_build]
[![Coverage][badge_coverage]][link_coverage]
[![Go Report][badge_goreport]][link_goreport]
[![License][badge_license]][link_license]

This package provides file-based cache implementation with entries expiration, checksum validation and another useful features.

Methods that interacts with file system uses mutexes, so, this cache implementation thread-safe, but performance in this case can be less than you might want.

## Installation and usage

The import path for the package is `github.com/tarampampam/go-filecache`.

To install it, run:

```bash
go get github.com/tarampampam/go-filecache
```

> API documentation can be [found here](https://godoc.org/github.com/tarampampam/go-filecache).

### Usage example

```go
package main

import (
    "bytes"
    "fmt"
    "time"

    filecache "github.com/tarampampam/go-filecache"
)

func main() {
    // Create new cache items pool
	// Note: Do NOT create new pool instance in goroutine - SHARE it instead
    pool := filecache.NewPool("/tmp")
    
    // Put data into cache pool with expiration time
    expiresAt := time.Now().Add(time.Minute * 10)
    if _, err := pool.Put("foo", bytes.NewBuffer([]byte("foo data")), expiresAt); err != nil {
        panic(err)
    }
    
    // Put data without expiration time
    if _, err := pool.PutForever("bar", bytes.NewBuffer([]byte("bar data"))); err != nil {
        panic(err)
    }

    // Define buffer for cached data reading
    buf := bytes.NewBuffer([]byte{})

    // Read data using reader
    if err := pool.GetItem("foo").Get(buf); err != nil {
        panic(err)
    }

    fmt.Println(buf) // "foo data"
}
```

Simple benchmark _(put value, get value, read into buffer and set new value)_ results:

```
goos: linux
goarch: amd64
pkg: github.com/tarampampam/go-filecache
BenchmarkSetAndGet-8       10000            611470 ns/op           40836 B/op         84 allocs/op
```

### Testing

For application testing we use built-in golang testing feature and `docker-ce` + `docker-compose` as develop environment. So, just write into your terminal after repository cloning:

```shell
$ make test
```

Or execute benchmarks:

```shell
$ make gobench
```

## Changelog

[![Release date][badge_release_date]][link_releases]
[![Commits since latest release][badge_commits_since_release]][link_commits]

Changes log can be [found here][link_changes_log].

## Support

[![Issues][badge_issues]][link_issues]
[![Issues][badge_pulls]][link_pulls]

If you will find any package errors, please, [make an issue][link_create_issue] in current repository.

## License

This is open-sourced software licensed under the [MIT License][link_license].

[badge_build]:https://img.shields.io/github/workflow/status/tarampampam/go-filecache/build?maxAge=30&logo=github
[badge_coverage]:https://img.shields.io/codecov/c/github/tarampampam/go-filecache/master.svg?maxAge=30
[badge_goreport]:https://goreportcard.com/badge/github.com/tarampampam/go-filecache
[badge_size_latest]:https://images.microbadger.com/badges/image/tarampampam/go-filecache.svg
[badge_release_version]:https://img.shields.io/github/release/tarampampam/go-filecache.svg?maxAge=30
[badge_language]:https://img.shields.io/github/go-mod/go-version/tarampampam/go-filecache?longCache=true
[badge_license]:https://img.shields.io/github/license/tarampampam/go-filecache.svg?longCache=true
[badge_release_date]:https://img.shields.io/github/release-date/tarampampam/go-filecache.svg?maxAge=180
[badge_commits_since_release]:https://img.shields.io/github/commits-since/tarampampam/go-filecache/latest.svg?maxAge=45
[badge_issues]:https://img.shields.io/github/issues/tarampampam/go-filecache.svg?maxAge=45
[badge_pulls]:https://img.shields.io/github/issues-pr/tarampampam/go-filecache.svg?maxAge=45
[link_goreport]:https://goreportcard.com/report/github.com/tarampampam/go-filecache

[link_coverage]:https://codecov.io/gh/tarampampam/go-filecache
[link_build]:https://github.com/tarampampam/go-filecache/actions
[link_license]:https://github.com/tarampampam/go-filecache/blob/master/LICENSE
[link_releases]:https://github.com/tarampampam/go-filecache/releases
[link_commits]:https://github.com/tarampampam/go-filecache/commits
[link_changes_log]:https://github.com/tarampampam/go-filecache/blob/master/CHANGELOG.md
[link_issues]:https://github.com/tarampampam/go-filecache/issues
[link_create_issue]:https://github.com/tarampampam/go-filecache/issues/new/choose
[link_pulls]:https://github.com/tarampampam/go-filecache/pulls


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog][keepachangelog] and this project adheres to [Semantic Versioning][semver].

## v1.0.2

### Fixed

- Cached file name length

## v1.0.1

### Fixed

- Cached file name generation (`sha1` changed to `md5`)

## v1.0.0

### Added

- First release

[keepachangelog]:https://keepachangelog.com/en/1.0.0/
[semver]:https://semver.org/spec/v2.0.0.html


item_test.go
package filecache

import (
	"bytes"
	"path/filepath"
	"strings"
	"sync"
	"testing"
	"time"
)

func TestItem_GetAndSet(t *testing.T) {
	t.Parallel()

	tmpDir := createTempDir(t)
	defer removeTempDir(t, tmpDir)

	key := "test-key"
	content := []byte(strings.Repeat("foo ", 32))
	item := newItem(NewPool(tmpDir), key)

	if err := item.Set(bytes.NewBuffer(content)); err != nil {
		t.Errorf("Got unexpected error on data SET: %v", err)
	}

	buf := bytes.NewBuffer([]byte{})
	if err := item.Get(buf); err != nil {
		t.Errorf("Got unexpected error on data GET: %v", err)
	}

	if !bytes.Equal(buf.Bytes(), content) {
		t.Errorf("Got unexpected content from cache item. Want: %v, got: %v", content, buf.Bytes())
	}
}

func TestItem_GetAndSetConcurrent(t *testing.T) { // nolint:gocyclo
	t.Parallel()

	tmpDir := createTempDir(t)
	defer removeTempDir(t, tmpDir)

	tests := []struct {
		name     string
		giveItem *Item
		setup    func(t *testing.T, item *Item)
	}{
		{
			name:     "Default set and set concurrent",
			giveItem: newItem(NewPool(tmpDir), strings.Repeat("foo", 64)),
			setup: func(t *testing.T, item *Item) {
				// setup basic state
				if err := item.Set(bytes.NewBuffer([]byte(strings.Repeat("x", 32)))); err != nil {
					t.Errorf("Got unexpected error on data SET: %v", err)
				}
				if err := item.SetExpiresAt(time.Now().Add(time.Second * 10)); err != nil {
					t.Errorf("Got unexpected error on set expiring time: %v", err)
				}
			},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			wg := sync.WaitGroup{}

			tt.setup(t, tt.giveItem)

			// start "Get" go routines
			for i := 0; i < 256; i++ {
				wg.Add(1)
				go func(item *Item) {
					defer wg.Done()
					item.GetKey()
					if err := item.Get(bytes.NewBuffer([]byte{})); err != nil {
						t.Errorf("Got unexpected error on data GET: %v", err)
					}
					if key := item.GetKey(); key != tt.giveItem.key {
						t.Errorf("Wrong key returged. Want: %s, got: %s", tt.giveItem.key, key)
					}
					if expTime := item.ExpiresAt(); expTime == nil {
						t.Error("Expiration time was not returned")
					}
					if _, err := item.IsExpired(); err != nil {
						t.Errorf("Got unexpected error on expiring checking: %v", err)
					}
				}(tt.giveItem)
			}

			// start "Set" go routines
			for i := 0; i < 256; i++ {
				wg.Add(1)
				go func(item *Item) {
					defer wg.Done()
					if err := item.Set(bytes.NewBuffer([]byte(strings.Repeat("z", 32)))); err != nil {
						t.Errorf("Got unexpected error on data SET: %v", err)
					}
					if key := item.GetKey(); key != tt.giveItem.key {
						t.Errorf("Wrong key returged. Want: %s, got: %s", tt.giveItem.key, key)
					}
					if err := item.SetExpiresAt(time.Now().Add(time.Second * 10)); err != nil {
						t.Errorf("Got unexpected error on set expiring time: %v", err)
					}
				}(tt.giveItem)
			}

			wg.Wait()
		})
	}
}

func TestItem_ExpiringGetSetAndCheck(t *testing.T) {
	t.Parallel()

	tmpDir := createTempDir(t)
	defer removeTempDir(t, tmpDir)

	item := newItem(NewPool(tmpDir), "a")

	if ok, isExpErr := item.IsExpired(); ok {
		t.Errorf("Just created item cannot be expirered")
	} else if isExpErr == nil {
		t.Errorf("Expected error on expirind checking was not returned")
	}

	expiresAt := time.Now().Add(3 * time.Millisecond)

	if err := item.SetExpiresAt(expiresAt); err != nil {
		t.Errorf("Unexpected error on expirind set: %v", err)
	}

	// wait for expiring
	time.Sleep(4 * time.Millisecond)

	if ok, isExpErr := item.IsExpired(); !ok {
		t.Error("Expired must return 'true' on `IsExpired` calling")
	} else if isExpErr != nil {
		t.Errorf("Unexpected error on expirind checking: %v", isExpErr)
	}

	if item.ExpiresAt().Unix() != expiresAt.Unix() {
		t.Errorf("Wrong `ExpiredAt` result. Want %v, got: %v", expiresAt, item.ExpiresAt())
	}
}

func TestItem_GetFilePath(t *testing.T) {
	t.Parallel()

	tmpDir := createTempDir(t)
	defer removeTempDir(t, tmpDir)

	item := newItem(NewPool(tmpDir), strings.Repeat("foo", 512))

	if !strings.HasSuffix(item.GetFilePath(), ".cache") {
		t.Errorf("Expected postfix [%s] was not found in %s", ".cache", item.GetFilePath())
	}

	if len(filepath.Base(item.GetFilePath())) > 32+len(".cache") {
		t.Errorf("Too long cache item file name: %s", filepath.Base(item.GetFilePath()))
	}
}


pool_test.go
package filecache

import (
	"bytes"
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"testing"
	"time"
)

func BenchmarkSetAndGet(b *testing.B) {
	tmpDir, _ := ioutil.TempDir("", "test-")
	defer func(b *testing.B) {
		if err := os.RemoveAll(tmpDir); err != nil {
			b.Fatal(err)
		}
	}(b)

	pool := NewPool(tmpDir)

	for i := 0; i < b.N; i++ {
		key := fmt.Sprintf("%s_%d", "test_key", i)

		if _, err := pool.PutForever(key, bytes.NewBuffer([]byte(strings.Repeat("x", i)))); err != nil {
			b.Fatal(err)
		}

		item := pool.GetItem(key)

		if err := item.Get(bytes.NewBuffer([]byte{})); err != nil {
			b.Fatal(err)
		}

		if err := item.Set(bytes.NewBuffer([]byte(strings.Repeat("z", i)))); err != nil {
			b.Fatal(err)
		}
	}
}

func TestNewPool(t *testing.T) {
	t.Parallel()

	giveGirPath := "foo"
	pool := NewPool(giveGirPath)

	if dirPath := pool.GetDirPath(); dirPath != "foo" {
		t.Errorf("Wrong directory path is set. Want: %s, got: %s", giveGirPath, dirPath)
	}
}

func TestPool_Clear(t *testing.T) {
	t.Parallel()

	tmpDir := createTempDir(t)
	defer removeTempDir(t, tmpDir)

	// Put "non-cache" file into temporary directory
	extraFilePath := filepath.Join(tmpDir, "extra_data")
	extraFile, _ := os.Create(extraFilePath)
	if err := extraFile.Close(); err != nil {
		panic(err)
	}

	// check for "non-cache" file exists
	if info, err := os.Stat(extraFilePath); err != nil || !info.Mode().IsRegular() {
		t.Fatalf("Cannot create non-cache file by path: %s", extraFilePath)
	}

	pool := NewPool(tmpDir)

	tests := []struct {
		name string
		data []byte
	}{
		{name: "foo", data: []byte("foo")},
		{name: "bar", data: []byte("bar")},
	}

	// Write cache items
	for _, tt := range tests {
		if _, err := pool.PutForever(tt.name, bytes.NewBuffer(tt.data)); err != nil {
			t.Error(err)
		}
	}

	// Check for items is exists
	for _, tt := range tests {
		if pool.HasItem(tt.name) != true {
			t.Errorf("Got `false` for just created cache item")
		}
	}

	// Make clear
	if result, clearErr := pool.Clear(); result != true || clearErr != nil {
		t.Errorf("Clearing failed. Result is: %v, Error: %v", result, clearErr)
	}

	// Check for items is NOT exists
	for _, tt := range tests {
		if pool.HasItem(tt.name) != false {
			t.Errorf("Got non-`false` for deleted cache item named `%s`", tt.name)
		}
	}

	// check for "non-cache" still exists
	if info, err := os.Stat(extraFilePath); err != nil || !info.Mode().IsRegular() {
		t.Fatalf("Non-cache file by path %s does not exists", extraFilePath)
	}
}

func TestPool_DeleteItem(t *testing.T) {
	t.Parallel()

	tmpDir := createTempDir(t)
	defer removeTempDir(t, tmpDir)

	pool := NewPool(tmpDir)

	// Set two cache items
	if _, err := pool.PutForever("foo", bytes.NewBuffer([]byte("foo"))); err != nil {
		t.Error(err)
	}
	if _, err := pool.PutForever("bar", bytes.NewBuffer([]byte("bar"))); err != nil {
		t.Error(err)
	}

	// Delete one item
	if result, err := pool.DeleteItem("bar"); result != true || err != nil {
		t.Errorf("Error while item deleting. Result: %v, error: %v", result, err)
	}

	// Make checks
	if pool.HasItem("bar") != false {
		t.Errorf("Just deleted item must returns `false` on exists checking")
	}
	if pool.HasItem("foo") != true {
		t.Errorf("Previous item must steel exists")
	}
}

func TestPool_GetDirPath(t *testing.T) {
	t.Parallel()

	pool := NewPool("foo")

	if path := pool.GetDirPath(); path != "foo" {
		t.Errorf("Unepected dir path. Want: %s, got: %s", "foo", path)
	}
}

func TestPool_GetItem(t *testing.T) {
	t.Parallel()

	tmpDir := createTempDir(t)
	defer removeTempDir(t, tmpDir)

	pool := NewPool(tmpDir)

	tests := []struct {
		name string
		data []byte
	}{
		{name: "foo", data: []byte("foo")},
		{name: "bar", data: []byte("bar")},
	}

	// Write cache items
	for _, tt := range tests {
		if _, err := pool.PutForever(tt.name, bytes.NewBuffer(tt.data)); err != nil {
			t.Error(err)
		}
	}

	// Check for items is exists
	for _, tt := range tests {
		if pool.HasItem(tt.name) != true {
			t.Errorf("Got `false` for just created cache item")
		}
	}

	// Check for items is NOT exists
	for _, tt := range tests {
		buf := bytes.NewBuffer([]byte{})

		if err := pool.GetItem(tt.name).Get(buf); err != nil || !bytes.Equal(buf.Bytes(), tt.data) {
			t.Errorf("Got wrong content for %s. Want: %v, got: %v", tt.name, tt.data, buf.Bytes())
		}
	}
}

func TestPool_Put(t *testing.T) {
	t.Parallel()

	tmpDir := createTempDir(t)
	defer removeTempDir(t, tmpDir)

	pool := NewPool(tmpDir)

	// Set items with "expires at" data
	if _, err := pool.Put("foo", bytes.NewBuffer([]byte("foo")), time.Now().Add(time.Millisecond*100)); err != nil {
		t.Error(err)
	}
	if _, err := pool.Put("bar", bytes.NewBuffer([]byte("bar")), time.Now().Add(time.Millisecond*200)); err != nil {
		t.Error(err)
	}

	// Wait for some time
	time.Sleep(time.Millisecond * 101)

	// And then check availability
	if pool.HasItem("foo") != false {
		t.Errorf("Expired cache item must be not available")
	}
	if pool.HasItem("bar") != true {
		t.Errorf("Non-expired cache item must be available")
	}

	// Wait for some time again
	time.Sleep(time.Millisecond * 100)

	if pool.HasItem("bar") != false {
		t.Errorf("Expired cache item must be not available")
	}
}

func TestPool_PutForever(t *testing.T) {
	t.Parallel()

	tmpDir := createTempDir(t)
	defer removeTempDir(t, tmpDir)

	pool := NewPool(tmpDir)

	if _, err := pool.PutForever("foo", bytes.NewBuffer([]byte("foo"))); err != nil {
		t.Error(err)
	}

	// Wait for some time
	time.Sleep(time.Millisecond * 20)

	if pool.HasItem("foo") != true {
		t.Errorf("Never expired cache item must always be available")
	}
}

func TestPool_ConcurrentUsage(t *testing.T) {
	t.Parallel()

	tmpDir := createTempDir(t)
	defer removeTempDir(t, tmpDir)

	pool := NewPool(tmpDir)

	if _, err := pool.PutForever("foo", bytes.NewBuffer([]byte("foo"))); err != nil {
		t.Error(err)
	}

	wg := sync.WaitGroup{}

	for i := 0; i < 1024; i++ {
		wg.Add(2)
		go func(pool *Pool) {
			defer wg.Done()

			item := pool.GetItem("foo")

			if err := item.Get(bytes.NewBuffer([]byte{})); err != nil {
				t.Errorf("Got unexpected error on data GET: %v", err)
			}
		}(pool)
		go func(pool *Pool) {
			defer wg.Done()

			item := pool.GetItem("foo")

			if err := item.Set(bytes.NewBuffer([]byte(strings.Repeat("z", 32)))); err != nil {
				t.Errorf("Got unexpected error on data SET: %v", err)
			}
		}(pool)
	}

	wg.Wait()
}


usage_test.go
package filecache

import (
	"bytes"
	"testing"
	"time"
)

func TestUsageCreateAndSetCacheItem(t *testing.T) {
	t.Parallel()

	tmpDir := createTempDir(t)
	defer removeTempDir(t, tmpDir)

	content := []byte("bar")
	pool := NewPool(tmpDir)
	item, err := pool.Put("foo", bytes.NewBuffer(content), time.Now().Add(time.Second*10))

	if err != nil {
		t.Fatal(err)
	}

	if item.IsHit() != true {
		t.Errorf("IsHit() must return `true` for new cache item, but returns `false`")
	}

	getItem := pool.GetItem("foo")
	readBuffer := bytes.NewBuffer([]byte{})

	if getError := getItem.Get(readBuffer); getError != nil {
		t.Errorf("Unexpected error while gettind occured: %v", getError)
	}

	if !bytes.Equal(readBuffer.Bytes(), content) {
		t.Errorf("Got unexpected content from cache item. Want: %v, got: %v", content, readBuffer.Bytes())
	}
}


pool.go
package filecache

import (
	"io"
	"io/ioutil"
	"os"
	"path/filepath"
	"time"

	"github.com/tarampampam/go-filecache/file"
)

type Pool struct {
	dirPath string
}

// NewPool creates new cache items pool.
func NewPool(dirPath string) *Pool {
	return &Pool{
		dirPath: dirPath,
	}
}

// GetDirPath returns cache directory path.
func (pool *Pool) GetDirPath() string { return pool.dirPath }

// GetItem returns a Cache Item representing the specified key.
func (pool *Pool) GetItem(key string) CacheItem {
	item := newItem(pool, key)

	// Make check for exists and "is expired?"
	if item.IsHit() {
		if expired, _ := item.IsExpired(); expired {
			_, _ = pool.DeleteItem(key)
		}
	}

	return item
}

// HasItem confirms if the cache contains specified cache item.
func (pool *Pool) HasItem(key string) bool {
	return pool.GetItem(key).IsHit()
}

func (pool *Pool) walkOverCacheFiles(fn func(string, os.FileInfo)) error {
	files, err := ioutil.ReadDir(pool.dirPath)
	if err != nil {
		return err
	}

	for _, f := range files {
		path := filepath.Join(pool.dirPath, f.Name())
		cacheFile, err := file.OpenRead(path, DefaultItemFileSignature)

		// skip "wrong" or errored file
		if err != nil || cacheFile == nil {
			continue
		}

		// verify file signature and close file (closing error will be skipped)
		matched, _ := cacheFile.SignatureMatched()

		if closeErr := cacheFile.Close(); matched && closeErr == nil {
			// if all is ok - fall the func
			fn(path, f)
		}
	}

	return nil
}

// Clear deletes all items in the pool.
func (pool *Pool) Clear() (bool, error) {
	var lastErr error

	err := pool.walkOverCacheFiles(func(path string, _ os.FileInfo) {
		if rmErr := os.Remove(path); rmErr != nil {
			lastErr = rmErr
		}
	})

	if err != nil {
		return false, err
	}

	if lastErr != nil {
		return false, lastErr
	}

	return true, nil
}

// DeleteItem removes the item from the pool.
func (pool *Pool) DeleteItem(key string) (bool, error) {
	item := newItem(pool, key)

	if rmErr := os.Remove(item.GetFilePath()); rmErr != nil {
		return false, rmErr
	}

	return true, nil
}

// Put a cache item with expiring time.
func (pool *Pool) Put(key string, from io.Reader, expiresAt time.Time) (CacheItem, error) {
	item, putError := pool.PutForever(key, from)

	if putError != nil {
		return item, putError
	}

	if err := item.SetExpiresAt(expiresAt); err != nil {
		return item, err
	}

	return item, nil
}

// Put a cache item without expiring time.
func (pool *Pool) PutForever(key string, from io.Reader) (CacheItem, error) {
	item := newItem(pool, key)

	if err := item.Set(from); err != nil {
		return item, err
	}

	return item, nil
}


errors.go
package filecache

// Cache error types
type ErrorType uint8

const (
	ErrUnknown ErrorType = iota
	ErrFileOpening
	ErrFileReading
	ErrFileWriting
	ErrExpirationDataNotAvailable
)

type Error struct {
	// The type of error
	Type ErrorType

	// The error message
	Message string

	// Previous error (for unwrapping)
	previous error
}

// String for representing error message.
func (e ErrorType) String() string {
	switch e {
	case ErrUnknown:
		return "unknown"
	case ErrFileOpening:
		return "cannot open file"
	case ErrFileReading:
		return "cannot read file"
	case ErrFileWriting:
		return "cannot write file"
	case ErrExpirationDataNotAvailable:
		return "expiration data is not available"
	}

	return "unrecognized error type"
}

// Error returns the error's message.
func (e *Error) Error() string {
	return e.Message
}

// Unwrap returns previous error. If previous error does not exits, Unwrap returns nil.
func (e *Error) Unwrap() error {
	return e.previous
}

// Creates new error instance. Previous error can be nil.
func newError(tp ErrorType, message string, prev error) *Error {
	return &Error{Type: tp, Message: message, previous: prev}
}


file/file.go
// File implementation for using as a cache entry.
package file

import (
	"bytes"
	"crypto/sha1" //nolint:gosec
	"encoding/binary"
	"errors"
	"fmt"
	"hash"
	"io"
	"os"
	"time"
)

// Read/write buffer size in bytes
const rwBufferSize byte = 32

type (
	// File signature
	FSignature []byte

	// File field offset and length
	offset uint8
	length uint8

	// File field for storing "File signature" (special flag for osFile identification among other files)
	ffSignature struct {
		offset
		length
	}

	// File field for storing "Expires At" label (in unix timestamp format with milliseconds)
	ffExpiresAtUnixMs struct {
		offset
		length
	}

	// File field for storing data "hash sum" (in SHA1 format)
	ffDataSha1 struct {
		offset
		length
	}

	// Field for useful data
	ffData struct {
		offset
	}

	// Cache osFile representation (all offsets must be set manually on instance creation action)
	File struct {
		ffSignature
		ffExpiresAtUnixMs
		ffDataSha1
		ffData
		Signature FSignature
		osFile    *os.File  // osFile on filesystem
		hashing   hash.Hash // SHA1 "generator" (required for hash sum calculation)
	}
)

var DefaultSignature = FSignature("#/CACHE ") // 35, 47, 67, 65, 67, 72, 69, 32

// newFile creates new osFile instance.
func newFile(osFile *os.File, signature FSignature) *File {
	// setup default osFile type bytes slice
	if signature == nil {
		signature = DefaultSignature
	}

	// File block offsets are below:
	// +----------------+-----------------------+-----------------+------------+
	// | Signature 0..7 |    Meta Data 8..63    | DataSHA1 64..83 | Data 84..n |
	// +----------------+-----------------------+-----------------+------------+
	// |                | ExpiresAtUnixMs 8..15 |                 |            |
	// +----------------+-----------------------+-----------------+------------+
	// |                |    RESERVED 16..63    |                 |            |
	// +----------------+-----------------------+-----------------+------------+
	return &File{
		ffSignature: ffSignature{
			offset: 0,
			length: 8,
		},
		ffExpiresAtUnixMs: ffExpiresAtUnixMs{
			offset: 8,
			length: 8,
		},
		ffDataSha1: ffDataSha1{
			offset: 64,
			length: 20,
		},
		ffData: ffData{
			offset: 84,
		},
		Signature: signature,
		osFile:    osFile,
		hashing:   sha1.New(), //nolint:gosec
	}
}

// Create or truncates the named osFile. If the osFile already exists, it will be truncated. If the osFile does not exist,
// it is created with passed mode (permissions).
// signature can be omitted (nil) - in this case will be used default osFile signature.
// Important: osFile with signature and data hashsum will be created immediately.
func Create(name string, perm os.FileMode, signature FSignature) (*File, error) {
	f, openErr := os.OpenFile(name, os.O_RDWR|os.O_CREATE|os.O_TRUNC, perm)
	if openErr != nil {
		return nil, openErr
	}

	file := newFile(f, signature)

	// write osFile signature
	if err := file.setSignature(file.Signature); err != nil {
		return nil, err
	}

	// requires for hashsum init
	if err := file.SetData(bytes.NewBuffer([]byte{})); err != nil {
		return nil, err
	}

	return file, nil
}

// Open the named osFile for reading and writing. If successful, methods on the returned osFile can be used for
// reading and writing. If there is an error, it will be of type *os.PathError.
// signature can be omitted (nil) - in this case will be used default osFile signature.
func Open(name string, perm os.FileMode, signature FSignature) (*File, error) {
	return open(name, os.O_RDWR, perm, signature)
}

// OpenRead opens the named osFile for reading. If successful, methods on the returned osFile can be used for reading; the
// associated osFile descriptor has mode O_RDONLY. If there is an error, it will be of type *os.PathError.
// signature can be omitted (nil) - in this case will be used default osFile signature.
func OpenRead(name string, signature FSignature) (*File, error) {
	return open(name, os.O_RDONLY, 0, signature)
}

func open(name string, flag int, perm os.FileMode, signature FSignature) (*File, error) {
	f, err := os.OpenFile(name, flag, perm)
	if err != nil {
		return nil, err
	}

	return newFile(f, signature), nil
}

// Name returns the name of the osFile as presented to Open.
func (file *File) Name() string { return file.osFile.Name() }

// Close the File, rendering it unusable for I/O. On files that support SetDeadline, any pending I/O operations
// will be canceled and return immediately with an error.
// Close will return an error if it has already been called.
func (file *File) Close() error {
	return file.osFile.Close()
}

// SignatureMatched checks for osFile signature matching. Signature should be set on osFile creation. This function can
// helps you to detect files that created by current package.
func (file *File) SignatureMatched() (bool, error) {
	fType, err := file.getSignature()
	if err != nil {
		return false, err
	}

	return bytes.Equal(*fType, file.Signature), nil
}

// GetSignature of current osFile signature as a typed slice of a bytes.
func (file *File) GetSignature() (*FSignature, error) { return file.getSignature() }

// getSignature of current osFile signature as a typed slice of a bytes.
func (file *File) getSignature() (*FSignature, error) {
	buf := make(FSignature, file.ffSignature.length)

	if n, err := file.osFile.ReadAt(buf, int64(file.ffSignature.offset)); err != nil && err != io.EOF {
		return nil, err
	} else if l := len(buf); n != l {
		// limit length for too small reading results
		buf = buf[0:n]
	}

	return &buf, nil
}

// setSignature allows to use only bytes slice of signature with length defined in osFile structure.
func (file *File) setSignature(signature FSignature) error {
	if l := len(signature); l != int(file.ffSignature.length) {
		return fmt.Errorf("wrong signature length: required length: %d, passed: %d", file.ffSignature.length, l)
	}

	if n, err := file.osFile.WriteAt(signature, int64(file.ffSignature.offset)); err != nil {
		return err
	} else if n != len(signature) {
		return errors.New("wrong wrote bytes length")
	}

	return nil
}

// GetExpiresAt for current osFile (with milliseconds).
func (file *File) GetExpiresAt() (time.Time, error) {
	ms, err := file.getExpiresAtUnixMs()

	// check for "value was set?"
	if ms == 0 && err == nil {
		err = errors.New("value was not set")
	}

	return time.Unix(0, int64(ms*uint64(time.Millisecond))), err
}

// getExpiresAtUnixMs returns unsigned integer value with ExpiresAt in UNIX timestamp format in milliseconds.
func (file *File) getExpiresAtUnixMs() (uint64, error) {
	buf := make([]byte, file.ffExpiresAtUnixMs.length)

	if _, err := file.osFile.ReadAt(buf, int64(file.ffExpiresAtUnixMs.offset)); err != nil && err != io.EOF {
		return 0, err
	}

	return binary.LittleEndian.Uint64(buf), nil
}

// SetExpiresAt sets the expiring value.
func (file *File) SetExpiresAt(t time.Time) error {
	return file.setExpiresAtUnixMs(uint64(t.UnixNano() / int64(time.Millisecond)))
}

// setExpiresAtUnixMs sets the expiring time in milliseconds in osFile content.
func (file *File) setExpiresAtUnixMs(ts uint64) error {
	buf := make([]byte, file.ffExpiresAtUnixMs.length)

	// pack unsigned integer into slice of bytes
	binary.LittleEndian.PutUint64(buf, ts)

	if n, err := file.osFile.WriteAt(buf, int64(file.ffExpiresAtUnixMs.offset)); err != nil {
		return err
	} else if n != len(buf) {
		return errors.New("wrong wrote bytes length")
	}

	return nil
}

// setDataSHA1 sets data hashsum as s slice ob bytes. Hash length must be correct.
func (file *File) setDataSHA1(h []byte) error {
	if l := len(h); l != int(file.ffDataSha1.length) {
		return fmt.Errorf("wrong hash length: required length: %d, passed: %d", file.ffDataSha1.length, l)
	}

	if n, err := file.osFile.WriteAt(h, int64(file.ffDataSha1.offset)); err != nil {
		return err
	} else if n != len(h) {
		return errors.New("wrong wrote bytes length")
	}

	return nil
}

// GetDataHash returns osFile data hash.
func (file *File) GetDataHash() ([]byte, error) { return file.getDataSHA1() }

// getDataSHA1 returns osFile data hash.
func (file *File) getDataSHA1() ([]byte, error) {
	buf := make([]byte, file.ffDataSha1.length)

	if _, err := file.osFile.ReadAt(buf, int64(file.ffDataSha1.offset)); err != nil && err != io.EOF {
		return buf, err
	}

	return buf, nil
}

// SetData sets the osFile data (content will be read from the passed reader instance).
func (file *File) SetData(in io.Reader) error { return file.setData(in) }

// setData sets the osFile data (content will be read from the passed reader instance).
func (file *File) setData(in io.Reader) error {
	buf := make([]byte, rwBufferSize)
	off := int64(file.ffData.offset)
	file.hashing.Reset()

	for {
		// read part of input data
		n, err := in.Read(buf)
		if err != nil {
			if err != io.EOF {
				return err
			}
			break
		}

		// limit length for too small reading results
		if l := len(buf); n != l {
			buf = buf[0:n]
		}

		// write content into required position
		wroteBytes, writeErr := file.osFile.WriteAt(buf, off)
		if writeErr != nil {
			return writeErr
		}
		// write into "hashing" too for hash sum calculation
		if _, err := file.hashing.Write(buf); err != nil {
			return err
		}

		// move offset
		off += int64(wroteBytes)
	}

	if err := file.setDataSHA1(file.hashing.Sum(nil)); err != nil {
		return err
	}

	return nil
}

// GetData read osFile data and write it to the writer.
func (file *File) GetData(out io.Writer) error { return file.getData(out) }

// getData read osFile data and write it to the writer.
func (file *File) getData(out io.Writer) error {
	buf := make([]byte, rwBufferSize)
	off := uint64(file.ffData.offset)
	file.hashing.Reset()

	for {
		// read part of useful data
		n, readErr := file.osFile.ReadAt(buf, int64(off))

		// Ignore EOF here (will be checked later). In any another case - we will return an error immediately
		if readErr != nil && readErr != io.EOF {
			return readErr
		}

		// limit length for too small reading results
		if l := len(buf); n != l {
			buf = buf[0:n]
		}

		// write content into out writer
		wroteBytes, writeErr := out.Write(buf)
		if writeErr != nil {
			return writeErr
		}

		// write into "hashing" too for hash sum calculation
		if _, err := file.hashing.Write(buf); err != nil {
			return err
		}

		// move offset
		off += uint64(wroteBytes)

		if readErr == io.EOF {
			break
		}
	}

	// calculate just read data hash
	dataHash := file.hashing.Sum(nil)

	// get existing hash
	existsHash, hashErr := file.getDataSHA1()
	if hashErr != nil {
		return hashErr
	}

	// if hashes mismatched - data was broken
	if !bytes.Equal(dataHash, existsHash) {
		return fmt.Errorf("data hashes mismatched. required: %v, current: %v", existsHash, dataHash)
	}

	return nil
}
